File: algorithms/neetcode/algorithms/algorithms/src/main/java/com/mycompany/leetcode/hard/MinimumNumberOfTapsToOpenToWaterAGarden/minimum-number-of-taps-to-open-to-water-a-garden-chappy1.py

from typing import List

class S:
    def minTaps(self, n: int, ranges: List[int]) -> int:
        for i in range(n + 1):
            ranges[i] = [i - ranges[i], i + ranges[i]]
        
        ranges.sort(key = lambda x : x[0])
        ans = 1
        curr = prev = 0
        maxval = -1
        while prev < n + 1:
            while prev < n + 1 and ranges[prev][0] <= curr:
                maxval = max(maxval, ranges[prev][1])
                prev += 1
            if maxval >= n:
                return ans
            elif maxval == curr:
                return -1
            else:
                curr = maxval
                ans += 1
        return -1
    

n = 5
ranges = [3,4,1,1,0,0]
print(S().minTaps(n, ranges))

File: algorithms/neetcode/algorithms/algorithms/src/main/java/com/mycompany/leetcode/hard/MinimumNumberOfTapsToOpenToWaterAGarden/MinimumNumberOfTapsToOpenToWaterAGardenChappy1.java
package com.mycompany.leetcode.hard.MinimumNumberOfTapsToOpenToWaterAGarden;

public class MinimumNumberOfTapsToOpenToWaterAGardenChappy1 {
    static int minTaps(int n, int[] ranges) {
        for(int idx = 0; idx <= n; idx++) {
            int start = Math.max(0, idx - ranges[idx]);
            int end = Math.min(n, idx + ranges[idx]);
            ranges[start] = end - start;
        }

        if(ranges[0] == 0) return -1;
        int laddres = ranges[0];
        int stairs = ranges[0];
        int minJumps = 1;

        for(int idx = 1; idx <= n; idx++) {
            if(idx == n) return minJumps;

            if(idx + ranges[idx] > laddres) laddres = idx + ranges[idx];

            stairs--;
            if(stairs == 0) {
                if(laddres <= idx) return -1;

                stairs = laddres - idx;
                minJumps++;
            }
        }
        return -1;
    }

    public static void main(String[] args) {
        int n = 5;
        int[] ranges = {3,4,1,1,0,0};

        System.out.println(minTaps(n, ranges));
    }
}
